### ISSUE TITLE
ActiveRecord::AssociationTypeMismatch when persisting record by with reference to current_user instead of by id

### ISSUE THREAD

**Issue by liangmicha** [rails/rails#40690]
### Steps to reproduce

We have a controller with the following code:
```
class SomeController < ApiController
  def create
    Request.create(
      user: current_user,
      metadata: params[:metadata]
    )
  end
end
```

Here is our `models/request.rb` file:
```
class Request < ApplicationRecord
  belongs_to :user, optional: true
end
```

Whenever we send  `POST` request, we get the following stack trace in the `Request.create` line:
```
ActiveRecord::AssociationTypeMismatch (User(#172280) expected, got #<User id: "adeb2139-2318-4fb0-be6d-8d086b39dae1", email: "email@g.co", created_at: "2020-11-25 21:06:20", updated_at: "2020-11-25 21:08:26", username: "test_user", phone_number: "123456789", name: "test_user"> which is an instance of User(#24300)):
```

If we change the code to
```
  def create
    Request.create(
      user_id: current_user.id,
      metadata: params[:metadata]
    )
  end
```
or
```
  def create
    Request.create(
      user: User.find(current_user.id),
      metadata: params[:metadata]
    )
  end
```
the error no longer happens.

### Expected behavior

Expected behavior is we should be able to pass in `current_user` to a `user: ` instead of having to specify the id field when creating an ActiveRecord.

### System configuration
This can only be reproduced only on `development`. While it's on `production` or `test` we do not see any issues. Also, before we upgraded to rails 6, we were on `2.6.5` and rails `5.2.3` and we had no issues with that.
**Rails version**:
ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-linux]

**Ruby version**:
6.0.3.4

Gem file:
```
source "https://rubygems.org"
ruby "2.7.2"

gem "active_model_serializers"
gem "activerecord-import"
gem "administrate"
gem "aws-sdk-s3"
gem "barnes"
gem "bootsnap", require: false
gem "devise"
gem "devise-jwt"
gem "money-rails"
gem "pg"
gem "puma"
gem "pundit"
gem "rack-canonical-host"
gem "rack-cors"
gem "rails", "~> 6.0.3"
gem "recipient_interceptor"
gem "rspec_api_documentation"
gem "sassc-rails"
gem "soft_delete_rails"
gem "sprockets", ">= 3.0.0"
gem "time_for_a_boolean"
gem "title"
gem "tzinfo-data", platforms: %i[mingw x64_mingw mswin jruby]
gem "uglifier"
gem "will_paginate", "~> 3.1.0"

group :development do
  gem "listen"
  gem "rack-mini-profiler", require: false
  gem "rails-erd"
  gem "rubocop-performance"
  gem "rubocop-rails"
  gem "spring"
  gem "web-console"
end

group :development, :test do
  gem "awesome_print"
  gem "bundler-audit", ">= 0.5.0", require: false
  gem "dotenv-rails"
  gem "pry-byebug"
  gem "pry-rails"
  gem "rspec-rails", "~> 4.0.1", group: %i[development test]
end

group :test do
  gem "database_cleaner-active_record"
  gem "db-query-matchers"
  gem "formulaic"
  gem "launchy"
  gem "rubocop-faker"
  gem "simplecov", require: false
  gem "timecop"
  gem "webmock"
end

gem "suspenders", group: %i[development test]

gem "bourbon", ">= 5.0.1"
gem "bullet", group: %i[development test]
gem "capybara-selenium", group: :test
gem "delayed_job_active_record"
gem "factory_bot_rails", group: %i[development test production]
gem "faker", group: %i[development test production]
gem "neat", ">= 3.0.1"
gem "oj"
gem "rack-timeout", group: :production
gem "shoulda-matchers", group: :test
gem "spring-commands-rspec", group: :development
```

**Comment by louim** [comment]
Your problem looks a lot like this issue: https://github.com/rails/rails/issues/37591. You should check if the workaround suggested in [this post](https://github.com/rails/rails/issues/37591#issuecomment-648479317) fix your problem.

**Comment by liangmicha** [comment]
For now our workaround is to define a `workaround_current_user` in the base `ApplicationController` as follows:
```
def workaround_current_user
    User.find(current_user.id)
end
```

**Comment by dhh** [comment]
Please test against 6.1.0.rc1 and see if this is fixed there. Sounds like it re: #37591.

**Comment by rwojnarowski** [comment]
@liangmicha this might be related to this gem `devise-jwt`, there is issue about this: https://github.com/waiting-for-dev/devise-jwt/issues/183 there is even working fix for this issue:  https://github.com/waiting-for-dev/devise-jwt/pull/186 

But `devise-jwt` depends on: `dry-configurable`, and it appears that there is some regression with this gem: https://github.com/dry-rb/dry-configurable/pull/97 . And Using master branch from dry-configurable with fix: https://github.com/dry-rb/dry-configurable/pull/99 worked for me

**Comment by jcmorrow** [comment]
Coworker of OP here - I can confirm that on our code base this is fixed by upgrading to `6.1`! 🎉 

Thanks y'all!
